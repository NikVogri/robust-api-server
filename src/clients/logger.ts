import { Request, Response } from 'express';
import { Service } from 'typedi';
import winston from 'winston';
import { AppError } from '../error/AppError';

@Service()
export class Logger {
    private logger: winston.Logger;

    private custom = {
        levels: {
            error: 0,
            info: 2,
            http: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.logger = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'http',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'info',
                format: winston.format.combine(winston.format.simple(), winston.format.colorize({ all: true })),
            }),
            new winston.transports.Console({
                level: 'error',
                format: winston.format.combine(
                    // winston.format.timestamp(), TODO: add timestamp
                    // TODO: make sure error only gets logged once! remove default transports for error?
                    winston.format.json(),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error, metadata?: Record<string, unknown>) {
        this.logger.error(message, AppError.toStringifiableObject(error), metadata);
    }

    info(message: string, metadata?: Record<string, unknown>) {
        this.logger.info(message, metadata);
    }

    http(req: Request, res: Response, error?: AppError) {
        const log = this.buildHttpLog(req, res, error);
        this.logger.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: AppError): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            url: req.url,
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: req.query,
                body: req.body ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: {
                message: error?.message,
                stack: error?.stack,
                thrown: error?.thrownError ? AppError.toStringifiableObject(error.thrownError) : undefined,
            },
        });
    }
}
