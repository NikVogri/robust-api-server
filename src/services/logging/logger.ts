import winston from 'winston';
import { AppError } from '../../error/AppError';
import { singleton } from 'tsyringe';
import { Request, Response } from '../../types';

@singleton()
export class Logger {
    public readonly winston: winston.Logger;

    private readonly custom = {
        levels: {
            error: 0,
            info: 1,
            http: 2,
            verbose: 3,
        },
        colors: {
            error: 'red',
            info: 'yellow',
            http: 'gray',
            verbose: 'gray',
        },
    };

    constructor() {
        winston.addColors(this.custom.colors);
        winston.remove(winston.transports.Console);

        this.winston = winston.createLogger({
            transports: this.getTransports(),
            levels: this.custom.levels,
        });
    }

    private getTransports = (): winston.transports.ConsoleTransportInstance[] => {
        return [
            // Logs appear in STDOUT
            new winston.transports.Console({
                level: 'verbose',
                format: winston.format.combine(
                    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                    winston.format.printf(
                        (i) => `${i.level} ${i.timestamp}: ${i.message} ${JSON.stringify(i.stack) ?? ''}`,
                    ),
                    winston.format.colorize({ all: true }),
                ),
            }),
        ];
    };

    error(message: string, error: Error) {
        this.winston.error(message, AppError.toStringifiableObject(error));
    }

    info(message: string) {
        this.winston.info(message);
    }

    http(req: Request, res: Response, error?: Error) {
        const log = this.buildHttpLog(req, res, error);
        this.winston.http(log);
    }

    private buildHttpLog(req: Request, res: Response, error?: unknown): string {
        return JSON.stringify({
            success: res.statusCode.toString().startsWith('2'), // success status codes are 2xx
            originalUrl: req.originalUrl,
            path: `${req.method} ${req.url}`,
            // TODO: Can also add auth fields in the future
            // user: {}
            client: {
                ip: req.ip,
                hostname: req.hostname,
                version: req.httpVersion,
                userAgent: req.headers['user-agent'],
            },
            req: {
                query: Object.keys(req.query).length > 0 ? req.query : undefined,
                params: Object.keys(req.params).length > 0 ? req.params : undefined,
                body: Object.keys(req.body).length > 0 ? Object.keys(req.body) : undefined,
                // do not log request inputs as they might be too large or contain GDPR data
            },
            res: {
                statusCode: res.statusCode,
                // do not log resposes as they might be too large or contain GDPR data
            },
            error: this.buildErrorObject(error),
        });
    }

    private buildErrorObject(error?: unknown) {
        if (!error || !(error instanceof Error)) return undefined;

        const output: Record<string, unknown> = {
            message: error?.message,
        };

        if (error instanceof AppError && error.thrownError) {
            output['thrown'] = AppError.toStringifiableObject(error.thrownError);
        }

        return output;
    }
}
