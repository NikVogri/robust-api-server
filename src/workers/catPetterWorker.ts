import { Job } from 'bullmq';
import { Logger } from '../clients/logger';
import { container } from 'tsyringe';

export const catPetterWorker = async (job: Job) => {
    const logger = container.resolve(Logger);

    logger.info('Started to pet a cat ' + job.data.name);
    await new Promise((r) => setTimeout(r, 100));
    logger.info('Completed petting cat ' + job.data.name);

    return true;
};
