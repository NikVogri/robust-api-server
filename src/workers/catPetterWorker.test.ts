import { describe, expect, it, vi } from 'vitest';
import { Logger } from '../services/logging/logger';
import { Job } from 'bullmq';
import { catPetterWorker } from './catPetterWorker';
import { container } from 'tsyringe';

vi.mock('../services/logging/logger', () => {
    const Logger = vi.fn(() => ({
        info: vi.fn(),
    }));

    return { Logger };
});

describe('Test catPetterWorker', () => {
    container.registerInstance(Logger, new Logger());
    const logger = container.resolve(Logger);

    it('should pet the cat', async () => {
        expect.assertions(3);

        vi.spyOn(logger, 'info').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return undefined;
        });

        vi.spyOn(logger, 'info').mockImplementationOnce((...args) => {
            expect(args).toMatchSnapshot();
            return undefined;
        });

        const job = {
            data: {
                name: 'Bob',
            },
        } as Job;

        const res = await catPetterWorker(job);
        expect(res).toMatchSnapshot();
    });
});
