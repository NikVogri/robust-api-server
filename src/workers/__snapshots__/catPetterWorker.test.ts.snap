// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test catPetterWorker > should pet the cat 1`] = `
[
  "Started to pet a cat Bob",
]
`;

exports[`Test catPetterWorker > should pet the cat 2`] = `
[
  "Completed petting cat Bob",
]
`;

exports[`Test catPetterWorker > should pet the cat 3`] = `true`;
